README.md
# Bottom Sheet Expandable Bar

This package create a bottom navigation bar with the capability to show a bottom sheet.

<img src="https://raw.githubusercontent.com/ajomuch92/flutter-bottom-sheet-expandable-bar/master/assets/demo.gif" width="200" height="429"/>

## Instalation
Include `bottom_sheet_expandable_bar` in your `pubspec.yaml` file:

```yaml
dependencies:
  flutter:
    sdk: flutter
  bottom_sheet_expandable_bar: version
```

## Usage

To use this package, just import it into your file and enjoy it.

```dart
import 'package:bottom_sheet_expandable_bar/bottom_sheet_bar_icon.dart';
import 'package:bottom_sheet_expandable_bar/bottom_sheet_expandable_bar.dart';

...

bottomSheet: BottomBarSheet(
  children: [
    BottomSheetBarIcon(
      icon: Icon(Icons.home),
      color: Colors.redAccent, 
      onTap: (){
      },
    ),
    BottomSheetBarIcon(
      icon: Icon(Icons.person),
      color: Colors.blueAccent, 
      onTap: (){
      },
    ),
    BottomSheetBarIcon(
      icon: Icon(Icons.edit),
      color: Colors.blue[800], 
      onTap: (){
      },
    ),
    BottomSheetBarIcon(
      icon: Icon(Icons.star),
      color: Colors.orangeAccent, 
      onTap: (){
      },
    ),
  ],
),

...

```

IMPORTANT: Add this line to your theme to avoid an unexpected behaviour

```dart
theme: ThemeData(
  /// Add this line
  bottomSheetTheme: BottomSheetThemeData(backgroundColor: Colors.transparent),
),

```

### BottomSheetBarIcon
This widget allow to generate a dynamic icon

**Properties**

|  Name | Description   | Required   | Default   |
| ------------ | ------------ | ------------ | ------------ |
| icon  | Widget to put as button icon | False   |   |
| color  | Color to indicate icon color | False   |  Colors.black |
| onTap  | Function to handle button on tap | True   |   |

### BottomBarSheet
This widget allow to generate a bottom bar with the capability to expanto to a bottom sheet

**Properties**

|  Name | Description   | Required   | Default   |
| ------------ | ------------ | ------------ | ------------ |
| children  | List of BottomSheetBarIcon widget to show icons within the bar | False   |   |
| buttonPosition  | ButtonBottomBarPosition to indicate expandable button position (center, end) |  False  | ButtonBottomBarPosition.center   |
| backgroundBarColor  | Color to indicate bottom bar background color | False   |  Colors.white |
| backgroundColor  | Color to indicate bottom sheet color | False   | Colors.white  |
| showExpandableButton  | Boolean value to indicate when to show or hide expandable button | False   |  False |
| innerChild  | Widget to show into the bottom sheet | False   |   |
| bottomRadius  | Double value to indicate the corners radius for the bottom bar | False   |  50.0 |
| bottomBarHeight  | Double value to indicate the bottom bar height | False   |  60.0 |
| bottomBarWidth  | Double value to indicate the bottom bar height | False   | Screen width * 0.9 |
| duration  | Duration between show or hide the bottom sheet | False   | 250 milliseconds |
| bottomSheetHeight  | Double value to indicate the bottom sheet height | False   | Screen height * 0.75  |
| iconExpand  | Icon to show into the expandable button | False   | Icon(Icons.navigation) |
| iconColor  | Expandable button color | False   | Colors.green |
| onClose  | Function to call when bottom sheet is closed | False   | (){}  |
| currentIndex  | Integer value to indicatet which tab is selected | False   | 0  |
| curve  | Curve to use on enter bottom sheet animation | False   |  Curves.ease |

### Expandable button at center

![](https://github.com/ajomuch92/flutter-bottom-sheet-expandable-bar/blob/master/assets/Screenshot_1.jpg)

### Expandable button at end

![](https://github.com/ajomuch92/flutter-bottom-sheet-expandable-bar/blob/master/assets/Screenshot_2.jpg)

### As a simple navigation bar

![](https://github.com/ajomuch92/flutter-bottom-sheet-expandable-bar/blob/master/assets/Screenshot_3.jpg)

CHANGELOG.md
<!-- DateFormat DD/MM/YYYY -->
## [0.0.1] - First release

## [0.1.0] - Second release (null safety)

## [0.1.1] - Third release

## [0.1.2] - 4th release (24/07/2021)

test/bottom_sheet_expandable_bar_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:bottom_sheet_expandable_bar/bottom_sheet_expandable_bar.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
    expect(() => calculator.addOne(null), throwsNoSuchMethodError);
  });
}


example/lib/main.dart
import 'package:bottom_sheet_expandable_bar/bottom_sheet_bar_icon.dart';
import 'package:bottom_sheet_expandable_bar/bottom_sheet_expandable_bar.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Bottom Bar Sheet Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
        bottomSheetTheme: BottomSheetThemeData(backgroundColor: Colors.transparent),
      ),
      debugShowCheckedModeBanner: false,
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key}) : super(key: key);

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _currentIndex = 0;
  
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: new AppBar(
        title: const Text('Bottom Bar Sheet Demo'),
        backgroundColor: Colors.blueGrey.shade900,
        centerTitle: true,
      ),
      body: Container(
        color: Colors.blueGrey.shade800,
        child: Center(
          child: Text('Demo', style: TextStyle(color: Colors.white),),
        ),
      ),
      bottomSheet: BottomBarSheet(
        buttonPosition: ButtonBottomBarPosition.center,
        innerChild: Center(child: Text('Bottom sheet'),),
        currentIndex: _currentIndex,
        curve: Curves.bounceIn,
        backgroundBarColor: Colors.grey,
        backgroundColor: Colors.yellowAccent,
        children: [
          BottomSheetBarIcon(
            icon: Icon(Icons.home),
            color: Colors.redAccent, 
            onTap: (){
              setState(() {
                _currentIndex = 0;
              });
            },
          ),
          BottomSheetBarIcon(
            icon: Icon(Icons.person),
            color: Colors.blueAccent, 
            onTap: (){
              setState(() {
                _currentIndex = 1;
              });
            },
          ),
          BottomSheetBarIcon(
            icon: Icon(Icons.edit),
            color: Colors.blue[800], 
            onTap: (){
              setState(() {
                _currentIndex = 2;
              });
            },
          ),
          BottomSheetBarIcon(
            icon: Icon(Icons.star),
            color: Colors.orangeAccent, 
            onTap: (){
              setState(() {
                _currentIndex = 3;
              });
            },
          ),
        ],
      ),
    );
  }
}


lib/bottom_sheet_bar_icon.dart
library bottom_sheet_expandable_bar;

import 'package:flutter/material.dart';

/// Class to generate the icons widget for bottom bar
class BottomSheetBarIcon extends StatelessWidget {
  /// Icon color, when is not active appers with opacity
  final Color? color;

  /// Icon to show within the widget
  final Icon? icon;

  /// Function trigger when pressed the icon
  final Function()? onTap;

  /// Boolean value to indicate when an icon is active
  final bool? isActive;

  /// icon: Icon color, when is not active appers with opacity
  /// color: Icon to show within the widget
  /// onTap: Function trigger when pressed the icon
  const BottomSheetBarIcon(
      {Key? key,
      @required this.icon,
      this.onTap,
      this.color = Colors.black,
      this.isActive = false})
      : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Column(
      mainAxisAlignment: MainAxisAlignment.center,
      children: [
        FloatingActionButton(
          child: this.icon,
          backgroundColor:
              this.color?.withOpacity(this.isActive == true ? 1 : 0.5),
          mini: true,
          elevation: 0.0,
          onPressed: this.onTap,
        ),
        AnimatedContainer(
          width: this.isActive == true ? 10.0 : 0.0,
          height: this.isActive == true ? 5.0 : 0.0,
          decoration: BoxDecoration(
            color: this.isActive == true ? this.color : Colors.transparent,
            borderRadius: BorderRadius.circular(5.0),
          ),
          duration: const Duration(milliseconds: 500),
          curve: Curves.fastOutSlowIn,
        )
      ],
    );
  }
}


lib/bottom_sheet_expandable_bar.dart
library bottom_sheet_expandable_bar;

import 'package:flutter/material.dart';
import 'package:bottom_sheet_expandable_bar/bottom_sheet_bar_icon.dart';

/// enum for expandable button position
enum ButtonBottomBarPosition { center, end }

/// Class to generate the bottom bar sheet widget
class BottomBarSheet extends StatefulWidget {
  /// List of BottomSheetBarIcon to show whitin the widget
  final List<BottomSheetBarIcon> children;

  /// Value to indicate the position of the expandable button (center, end)
  final ButtonBottomBarPosition? buttonPosition;

  /// Color for container when is expanded
  final Color? backgroundBarColor;

  /// Bottom bar calor
  final Color? backgroundColor;

  /// Boolean value to show the expandable button
  final bool showExpandableButton;

  /// Widget to render into the bottom sheet
  final Widget? innerChild;

  /// Double value to indicate when radius for corner on bottom bar
  final double bottomRadius;

  /// Value to indicate the bottom bar height
  final double? bottomBarHeight;

  /// Value to indicate the bottom bar width
  final double? bottomBarWidth;

  /// Indicate the duration when expand bottom sheet
  final Duration? duration;

  /// Double value to indicate the bottom sheet height
  final double? bottomSheetHeight;

  /// Icon to show into expandable button
  final Widget? iconExpand;

  /// Color for expandable button
  final Color? iconColor;

  /// Function trigger when close bottom sheet
  final Function()? onClose;

  /// Value to indicate what icon tab is selected
  final int? currentIndex;

  /// Curve to use on animation
  final Curve curve;

  /// children: List of BottomSheetBarIcon to show whitin the widget
  /// buttonPosition: Value to indicate the position of the expandable button (center, end)
  /// backgroundColor: Color for container when is expanded
  /// backgroundBarColor: Bottom bar calor
  /// showExpandableButton: Boolean value to show the expandable button
  /// innerChild: Widget to render into the bottom sheet
  /// bottomRadius: Double value to indicate when radius for corner on bottom bar
  /// bottomBarHeight: Value to indicate the bottom bar height
  /// bottomBarWidth: Value to indicate the bottom bar width
  /// duration: Indicate the duration when expand bottom sheet
  /// bottomSheetHeight: Double value to indicate the bottom sheet height
  /// iconExpand: Icon to show into expandable button
  /// iconColor: Color for expandable button
  /// onClose: Function trigger when close bottom sheet
  /// currentIndex: Value to indicate what icon tab is selected
  /// curve: Curve to use on animation
  BottomBarSheet({
    Key? key,
    this.children = const [],
    this.buttonPosition = ButtonBottomBarPosition.center,
    this.backgroundColor = Colors.white,
    this.backgroundBarColor = Colors.white,
    this.showExpandableButton = true,
    this.innerChild,
    this.bottomRadius = 50.0,
    this.bottomBarHeight = 60.0,
    this.bottomBarWidth,
    this.duration = const Duration(milliseconds: 250),
    this.bottomSheetHeight,
    this.iconExpand = const Icon(Icons.navigation),
    this.iconColor = Colors.green,
    this.onClose,
    this.currentIndex,
    this.curve = Curves.ease,
  }) : super(key: key);

  @override
  _BottomBarSheetState createState() => _BottomBarSheetState();
}

class _BottomBarSheetState extends State<BottomBarSheet> {
  /// Boolean value to toggle first or second child on AnimatedCrossFade
  bool _showSecond = false;

  List<Widget> _getChildren() {
    List<Widget> _children = [];
    int middle = (widget.children.length / 2).ceil();
    int _index = 0;
    for (BottomSheetBarIcon el in widget.children) {
      int i = widget.children.indexOf(el);
      BottomSheetBarIcon _icon = BottomSheetBarIcon(
          icon: el.icon,
          onTap: () {
            el.onTap!();
          },
          isActive: widget.currentIndex == i,
          color: el.color);
      _children.add(_icon);
      _index++;
      if (_index == middle &&
          widget.buttonPosition == ButtonBottomBarPosition.center &&
          widget.showExpandableButton) {
        _children.add(_getShowButton());
      }
    }
    if (widget.buttonPosition == ButtonBottomBarPosition.end &&
        widget.showExpandableButton) {
      _children.add(_getShowButton());
    }
    return _children;
  }

  Widget _getShowButton() {
    return FloatingActionButton(
      child: widget.iconExpand,
      elevation: 0.0,
      backgroundColor: widget.iconColor,
      onPressed: () => setState(() => _showSecond = true),
    );
  }

  @override
  Widget build(BuildContext context) {
    Size _size = MediaQuery.of(context).size;

    return BottomSheet(
      /// when on close is null, take an empty function
      onClosing: widget.onClose ?? () {},
      builder: (BuildContext context) => AnimatedContainer(
        margin: EdgeInsets.all(20),
        decoration: BoxDecoration(
            color: !_showSecond
                ? widget.backgroundBarColor
                : widget.backgroundColor,
            borderRadius: BorderRadius.circular(30)),
        child: AnimatedCrossFade(
            firstChild: Container(
              height: widget.bottomBarHeight,
              width: widget.bottomBarWidth ?? (_size.width * 0.9),
              decoration: BoxDecoration(
                borderRadius:
                    BorderRadius.all(Radius.circular(widget.bottomRadius)),
                // color: widget.backgroundBarColor,
              ),
              child: Row(
                mainAxisAlignment: MainAxisAlignment.spaceAround,
                crossAxisAlignment: CrossAxisAlignment.center,
                children: _getChildren(),
              ),
            ),
            secondChild: Container(
                // color: widget.backgroundColor,
                constraints: BoxConstraints.expand(
                  height: widget.bottomSheetHeight ?? (_size.height * 0.75),
                ),
                padding: EdgeInsets.all(20),
                child: Stack(
                  children: [
                    Container(
                      child: widget.innerChild,
                    ),
                    Positioned(
                        right: 0,
                        top: 0,
                        child: CloseButton(
                          color: Colors.redAccent,
                          onPressed: () {
                            setState(() => _showSecond = false);
                          },
                        ))
                  ],
                )),
            crossFadeState: _showSecond
                ? CrossFadeState.showSecond
                : CrossFadeState.showFirst,
            duration: widget.duration ?? Duration(milliseconds: 250)),
        duration: widget.duration ?? Duration(milliseconds: 250),
        curve: widget.curve,
      ),
    );
  }
}
